Q1. Define two variables, a and b, with values 12 and 5 respectively. Write a Python program to calculate and print their sum, difference, product, quotient, and modulus.

In [ ]:
#Define two variables, a and b, with values 12 and 5 respectively

a = 12 #defining variable a with value 12

b = 5 #defining variable b with value 5

print("Sum of a and b:",a+b) #adding a and b

print("Difference between a and b:",a-b) #subtracting b from a

print("Product of a and b:",a*b) #multiplying a and b

print("Quotient of a and b:",a/b) #dividing a by b

print("Modulus of a and b:",a%b) #finding modulus of a and b

Sum of a and b: 17
Difference between a and b: 7
Product of a and b: 60
Quotient of a and b: 2.4
Modulus of a and b: 2


Q2. Write a Python program to evaluate the expression (8 + 2) * (3 - 1) ** 2 / 2 and explain how the use of parentheses affects the result using comments.


In [ ]:
result = (8 + 2) * (3 - 1) ** 2 / 2

print("The result of the expression (8 + 2) * (3 - 1) ** 2 / 2 is:", result) # Print the result.

The result of the expression (8 + 2) * (3 - 1) ** 2 / 2 is: 20.0


### Explanation of how parentheses affect the result:

1. (8 + 2): Parentheses ensure that the addition is performed first, giving us 10.

2. (3 - 1): Parentheses ensure that the subtraction is performed next, giving us 2.

3. (3 - 1) ** 2: The result of the subtraction (2) is then squared, giving us 4.

4. (8 + 2) * (3 - 1) ** 2: This results in 10 * 4, which equals 40.

5. (8 + 2) * (3 - 1) ** 2 / 2: Finally, 40 is divided by 2, giving us the final result of 20.0.

Without parentheses, the order of operations (PEMDAS/BODMAS) would change the result.

Q3.Given three boolean variables, p = True, q = False, and r = True, write a Python program that evaluates and prints the result of the logical expressions p and q, p or q, and not r.

In [ ]:
# Given three boolean variables

p = True

q = False

r = True

# Evaluating and printing the result of the logical expressions

print ("The value of (p and q) is:", p and q) # printing the logical expression (p and q)

print ("The value of (p or q) is:", p or q) # printing the logical expression (p or q)

print ("The value of (not r) is:", not r) # printing the logical expression (not r)

The value of (p and q) is: False
The value of (p or q) is: True
The value of (not r) is: False


Q4. Write a Python program to evaluate the expression 10 + 2 * 3 / 4 - 5 ** 2 and print the result. Explain the order in which the operations are performed using comments.

In [ ]:
expression = 10 + 2 * 3 / 4 - 5 ** 2

print("The result of the expression 10 + 2 * 3 / 4 - 5 ** 2 is:", expression) # printing the given expression

The result of the expression 10 + 2 * 3 / 4 - 5 ** 2 is: -13.5


### Explanation of the order of operations:

1. Exponents : 5 ** 2 is calculated first, giving us 25.

2. Multiplication and Division (from left to right) :  

   * Multiplication - 2 * 3 is calculated next, giving us 6.

   * Division - 6 / 4 is then calculated, giving us 1.5.

3. Addition and Subtraction (from left to right) :

   * Addition - 10 + 1.5 is calculated next, giving us 11.5.

   * Subtraction - 11.5 - 25 is calculated last, giving us -13.5.

Q5. Explain the difference between logical operators (and, or, not) and bitwise operators (&, |, ~) with examples.

Logical operators (`and`, `or`, `not`) and bitwise operators (`&`, `|`, `~`) are used in programming for different purposes, although they can sometimes appear similar.

### Logical Operators:

Logical operators are used to combine or modify boolean values (True or False). They work at the level of boolean logic and are used in control flow statements.

### Bitwise Operators:

Bitwise operators perform operations on the individual bits of integer values. They are used in lower-level programming, such as systems programming, for direct manipulation of binary data.


In [ ]:
# creating sample variables to show difference between logical and bitwise operators

a = 20 # creating a sample variable a

b = 30 # creating a sample variable b

print(a > 50 and b > 50) # printing the logical ' and ' operation

print(a < 50 and b < 50) # printing the logical ' and ' operation

print(a & b) # printing the bitwise ' & ' operation

print(a > 50 or b < 50) # printing the logical ' or ' operation

print(a < 50 or b > 50) # printing the logical ' or ' operation

print(a | b)  # printing the bitwise ' or ' operation

print(not(a < 50)) # printing the logical ' not ' operation

print(not(a > 50))  # printing the logical ' not ' operation

print(~a) # printing the bitwise ' ~ ' operation

print(~b) # printing the bitwise ' ~ ' operation

False
True
20
True
True
30
False
True
-21
-31
